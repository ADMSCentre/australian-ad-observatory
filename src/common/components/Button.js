import React from 'react';
import PropTypes from 'prop-types';
import BButton from 'react-bootstrap/Button';

const openPage = to => {
  const base = 'webpage/index.html';
  const url = to ? `${base}#${to}` : base;
  const extensionUrl = chrome.runtime.getURL(url);

  /* istanbul ignore next */
  return () => {
    window.open(extensionUrl, '_blank');
  };
};

const Button = ({ type, label, to, variant, onClick }) => {
  const doClick = onClick || openPage(to);

  return type === 'link' ? (
    <a href="#" onClick={doClick}>
      {label}
    </a>
  ) : (
    <BButton variant={variant} block onClick={doClick}>
      {label}
    </BButton>
  );
};
Button.displayName = 'Button';

Button.propTypes = {
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
  type: PropTypes.oneOf(['link', 'button']),
  to: PropTypes.string, // you must have either a 'to' or 'onClick'
  onClick: PropTypes.func,
  variant: PropTypes.oneOf([
    'primary',
    'secondary',
    'success',
    'warning',
    'danger',
    'info',
    'light',
    'dark',
    'link'
  ])
};

Button.defaultProps = {
  type: 'button',
  to: null,
  onClick: null,
  variant: undefined
};

export default Button;
