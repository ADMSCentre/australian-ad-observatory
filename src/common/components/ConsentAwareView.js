import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import useConsent from 'common/hooks/useConsent';
import proceed from 'common/utils/proceed';

const ConsentAwareView = ({ children, whenGrantedIs, whenGrantedIsNot }) => {
  const { granted } = useConsent();

  return proceed(granted, whenGrantedIs, whenGrantedIsNot) ? <Fragment>{children}</Fragment> : null;
};
ConsentAwareView.displayName = 'Consent aware view';

ConsentAwareView.propTypes = {
  children: PropTypes.node.isRequired,
  whenGrantedIs: PropTypes.oneOf(['none', 'old', 'current']),
  whenGrantedIsNot: PropTypes.oneOf(['none', 'old', 'current'])
};

ConsentAwareView.defaultProps = {
  whenGrantedIs: null,
  whenGrantedIsNot: null
};

export default ConsentAwareView;
