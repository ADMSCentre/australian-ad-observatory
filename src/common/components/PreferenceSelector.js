import React from 'react';
import PropTypes from 'prop-types';
import usePreference from 'common/hooks/usePreference';
import I18n from 'common/i18n';

const buildOptions = options =>
  options.map(({ value, text }) => (
    <option key={value} value={value}>
      {text}
    </option>
  ));

const PreferenceSelector = ({ label, options, storageKey }) => {
  const { selected = '', savePreference } = usePreference(storageKey);

  return (
    <div className="pt-3">
      <label className="form-label">{label}</label>
      <select id={storageKey} className="form-control" onChange={savePreference} value={selected}>
        <option value="">{I18n('prefselector', 0)}...</option>
        {buildOptions(options)}
      </select>
    </div>
  );
};
PreferenceSelector.displayName = 'Preference selector';

PreferenceSelector.propTypes = {
  label: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired
    })
  ).isRequired,
  storageKey: PropTypes.string.isRequired
};

export default PreferenceSelector;
