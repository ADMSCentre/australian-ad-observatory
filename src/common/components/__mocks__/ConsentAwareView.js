/* eslint-disable react/prop-types */
import React, { Fragment } from 'react';
import proceed from 'common/utils/proceed';

let granted;

export const setGranted = gtd => {
  granted = gtd;
};

const ConsentAwareView = ({ children, whenGrantedIs, whenGrantedIsNot }) =>
  proceed(granted, whenGrantedIs, whenGrantedIsNot) ? <Fragment>{children}</Fragment> : null;

export default ConsentAwareView;
