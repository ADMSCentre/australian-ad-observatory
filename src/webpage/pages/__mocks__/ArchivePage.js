import React from 'react';

const ArchivePage = () => <h1>Archive Page goes here</h1>;

export default ArchivePage;
