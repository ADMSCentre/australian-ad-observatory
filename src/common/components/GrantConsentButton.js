import React from 'react';
import PropTypes from 'prop-types';
import Button from './Button';

import useConsent from 'common/hooks/useConsent';
import useNavigation from 'common/hooks/useNavigation';

const GrantConsentButton = ({ label, to }) => {
  const { granted, saveConsent } = useConsent();
  const { goto } = useNavigation();

  const saveAndGo = () => {
    saveConsent();
    if (to) goto(to);
  };

  return granted === null || granted === 'current' ? null : (
    <Button variant="primary" onClick={saveAndGo} label={label} />
  );
};
GrantConsentButton.displayName = 'Grant aware button';

GrantConsentButton.propTypes = {
  label: PropTypes.string.isRequired,
  to: PropTypes.string
};

GrantConsentButton.defaultProps = {
  to: null
};

export default GrantConsentButton;
