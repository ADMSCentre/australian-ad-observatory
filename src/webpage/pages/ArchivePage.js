import React, { Fragment } from 'react';
import ConsentAwareView from 'common/components/ConsentAwareView';
import Archive from '../components/Archive';
import I18n from 'common/i18n';

const ArchivePage = () => (
  <Fragment>
    <header>
      <h1>{I18n('archive', 0)}</h1>
    </header>
    <ConsentAwareView whenGrantedIs="none">
      <p className="introduction">
        You need to accept the terms and conditions before you can view your archive.
      </p>
      <a href="#terms">View and accept terms and conditions</a>
    </ConsentAwareView>
    <ConsentAwareView whenGrantedIs="old">
      <p className="introduction">
        You need to accept the updated terms and conditions before you can view your archive.
      </p>
      <a href="#terms">View and accept updated terms and conditions</a>
    </ConsentAwareView>
    <ConsentAwareView whenGrantedIs="current">
      <Archive />
    </ConsentAwareView>
  </Fragment>
);
ArchivePage.displayName = 'ArchivePage';

export default ArchivePage;
