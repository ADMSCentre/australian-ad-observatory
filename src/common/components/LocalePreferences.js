import React, { Fragment } from 'react';
import PreferenceSelector from './PreferenceSelector';
import { PREF_COUNTRY, PREF_LANGUAGE, PREF_AGE, PREF_GENDER } from 'common/keys';
import { COUNTRIES } from 'common/countries';
import { LANGUAGES } from 'common/languages';
import { AGES } from 'common/ages';
import { GENDERS } from 'common/genders';

const toArray = object => Object.keys(object).map(value => ({ value, text: object[value] }));

const LocalePreferences = () => (
  <Fragment>
    <PreferenceSelector
      options={toArray(COUNTRIES)}
      label="Your Country"
      storageKey={PREF_COUNTRY}
    />
    <PreferenceSelector
      options={toArray(LANGUAGES)}
      label="Your Language"
      storageKey={PREF_LANGUAGE}
    />
    <PreferenceSelector options={toArray(AGES)} label="Age" storageKey={PREF_AGE} />
    <PreferenceSelector options={toArray(GENDERS)} label="Gender" storageKey={PREF_GENDER} />
  </Fragment>
);
LocalePreferences.displayName = 'Locale preferences';

export default LocalePreferences;
