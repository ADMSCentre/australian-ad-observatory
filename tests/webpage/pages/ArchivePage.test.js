import React from 'react';
import renderer from 'react-test-renderer';
import ArchivePage from 'webpage/pages/ArchivePage';

import { setGranted } from 'common/components/ConsentAwareView';

jest.mock('common/components/ConsentAwareView');
jest.mock('webpage/components/Archive');

['none', 'old', 'current'].forEach(granted => {
  describe(`when consent is '${granted}'`, () => {
    let tree;

    beforeAll(() => {
      setGranted(granted);
      tree = renderer.create(<ArchivePage />);
    });

    it('rendered correctly', () => {
      expect(tree).toMatchSnapshot();
    });
  });
});
