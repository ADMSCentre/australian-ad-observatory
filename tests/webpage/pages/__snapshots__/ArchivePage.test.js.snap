// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when consent is 'current' rendered correctly 1`] = `
Array [
  <header>
    <h1>
      My Archive
    </h1>
  </header>,
  <div
    id="archive-list"
  >
    Archive stuff here
  </div>,
]
`;

exports[`when consent is 'none' rendered correctly 1`] = `
Array [
  <header>
    <h1>
      My Archive
    </h1>
  </header>,
  <p
    className="introduction"
  >
    You need to accept the terms and conditions before you can view your archive.
  </p>,
  <a
    href="#terms"
  >
    View and accept terms and conditions
  </a>,
]
`;

exports[`when consent is 'old' rendered correctly 1`] = `
Array [
  <header>
    <h1>
      My Archive
    </h1>
  </header>,
  <p
    className="introduction"
  >
    You need to accept the updated terms and conditions before you can view your archive.
  </p>,
  <a
    href="#terms"
  >
    View and accept updated terms and conditions
  </a>,
]
`;
